specgen/specgen.go

```diff
@@ -0,0 +1,518 @@
+// Package specgen provides a utility for generating utls.ClientHelloSpec.
+package specgen
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"reflect"
+
+	tls "github.com/getlantern/utls"
+)
+
+const (
+	// When we print byte slices, we print this many bytes per line.
+	bytesPerLine = 12
+
+	tlsRecordHeaderLen = 5
+)
+
+// WriteHelloSpec writes a utls.ClientHelloSpec corresponding to the input ClientHello.
+//
+// packagePrefix can be used to control the package qualifier on utls definitions. For example, if
+// packagePrefix is provided as "tls", then cipher suites will be specified like
+// "tls.TLS_AES_128_GCM_SHA256" in the output code. If packagePrefix is the empty string, then no
+// qualifier will be used. The example cipher suite would be printed as "TLS_AES_128_GCM_SHA256".
+func WriteHelloSpec(w io.Writer, clientHello []byte, packagePrefix string) error {
+	prefix := ""
+	if packagePrefix != "" {
+		prefix = packagePrefix + "."
+	}
+
+	helloSpec, err := tls.FingerprintClientHello(clientHello[tlsRecordHeaderLen:])
+	if err != nil {
+		return fmt.Errorf("failed to fingerprint hello: %w", err)
+	}
+
+	fmt.Fprintf(w, "%sClientHelloSpec{\n", prefix)
```

Comment:
You already but some effort into this, but this is a clear usecase for `text/template`, and it would read better.

Reply:
Ahh you're probably right.  That package is so powerful and I always forget it exists.  Hmm I'll think about switching.